torch/csrc/jit/ir/alias_analysis.cpp

```diff
@@ -878,6 +874,44 @@ void AliasDb::analyzeConservative(Node* node) {
   }
 }
 
+bool AliasDb::functionalNonEscapingListUse(const Use& use) const {
+  Node* n = use.user;
+  size_t offset = use.offset;
+  Value* container = n->inputs().at(offset);
+
+  // only consider aten op uses of lists
+  if (!container->type()->cast<ListType>()) {
+    return false;
+  }
+
+  /*
+  in the general case, we consider any Value that enters another container as
+  entering the heap, and thus aliasing all other heap values of the same type.
+  the advantage of this approach are:
+  - there are many composite list/container ops that would be tricky to
+  schematize if we did something more complicated
+  - limits the size of the AliasDb, because a container of size 10 only contains
+  1 memory dag element instead of 10
+  - we do not need to worry about adding contained elements to the wildcard set
+  when a container escapes the graph.
+  The downside of this approach is we are unable to handle the common case of a
+  list constructed and passed into an aten op. Here, optimize for a set of
+  common ops where the output does not alias the list or the list elements
+  */
+
+  switch (use.user->kind()) {
```

Comment:
What are your thoughts on how we should handle these in the long run? Ideally, this pass doesn't know much about specific ops (besides language primitives) and there is a structured way (alias annotations) for the builtin operator library to communicate aliasing/mutation. Do we feel like aliasing annotations are insufficiently expressive for what we want? Too hard to change?

Reply:
It's certainly possible for this to be supported in the future. One difficulty is that this is backwards-incompatible change; a schema like `aten::dict((Tensor, Tensor)[] inputs -> Dict(Tensor, Tensor)` doesn't currently annotate anything about the alias set of the contained types. We don't need to specify that the keys and values of the input list now are respectively in the keys and values of the Dict type because they were all already in the wildcard set to begin with. Maybe something like:
`aten::dict((Tensor(a->c), Tensor(b->d)[] inputs -> Dict(Tensor(c), Tensor(d))` or just `aten::dict((Tensor(a), Tensor(b))[] inputs -> Dict(Tensor(a), Tensor(b))`. I don't think this is necessarily blocking, it's just something we would have to consider... There are not that many of these ops, especially in userland. 

There are other implementation difficulties as well, and potentially compilation regressions. I remember something Densenet has a list which has 64 different append nodes, so now the contained elements go from 1 -> 64. 

The maintainability and consistency of alias analysis is obviously important, but at the end of a day it is a tool that enables other optimization, and not a goal in and of itself. I would generally like to avoid sinking too much time into improving alias analysis if doesn't buy us anything.

Reply:
@suo wdyt?